### ISSUE TITLE
RangeError when running on codeship platform

### ISSUE THREAD

**Issue by Mixolyde** [kevmoo/peanut.dart#88]
I am building this github repo with peanut
https://github.com/Mixolyde/blaseballstatus
And it works fine on my home windows laptop.
When I run peanut on CodeShip, a public CI/CD web app, I get the attached error.
![Screenshot_20200914-144749_Chrome](https://user-images.githubusercontent.com/639676/93214138-cb551e00-f732-11ea-9c6e-a70f0b94cebd.jpg)
I am sure this has something to do with their environment, but the error message isn't too helpful. Basically they spin up an empty linux box, clone your repo and then execute whatever shell commands you give it. The full build output might be viewable here: 

https://app.codeship.com/projects/408806/builds/48930950?pipeline=4e51b1cf-84a2-4057-b69b-ca1d74a4214e

Any insight on how to debug or diagnose this weird issue would be great.

**Comment by Mixolyde** [comment]
I ran this chunk of code on the codeship platform: https://github.com/Mixolyde/blaseballstatus/blob/master/tools/peanut_test.dart
And got this output:
Instance of 'GitDir'
HEAD

9b3ec6d62a8fd0c020596a2503b54d5c0d77e275 HEAD

[GitReference: HEAD  9b3ec6d62a8fd0c020596a2503b54d5c0d77e275]

So the git getBranch code is getting back "HEAD" instead of refs/heads/master, and crashing out.

**Comment by kevmoo** [comment]
The git code here is 10yrs old and likely makes assumptions about the structure of the local git config that don't hold on CodeShip – hrm...

**Comment by kevmoo** [comment]
Let me dig!

**Comment by kevmoo** [comment]
In progress - https://github.com/kevmoo/git/pull/21

**Comment by kevmoo** [comment]
If you reactivate `peanut` you should get the latest `pkg:git` – and everything *should* work 😄

**Comment by Mixolyde** [comment]
Thank you, thank you, thank you! It worked, now I just need to get codeship to publish the gh-pages branch upstream and my auto deployment will work!

**Comment by Mixolyde** [comment]
You might want to bump the pubspec for peanut to require the latest git package by default.

**Comment by kevmoo** [comment]
Given your the first and only person to hit this. I'm not too worried. By
default, global activate gets the latest dependencies, so it really doesn't
help much.

**Comment by Mixolyde** [comment]
So this may be related to running in the detached state, but on my deployment app peanut is working and passing, but when I try to push the remote branch, I get:
https://app.codeship.com/projects/408806/builds/48960635?pipeline=4c029e2f-c121-417f-9d49-12b2f66f0def

To github.com:Mixolyde/blaseballstatus.git
! [rejected]        gh-pages -> gh-pages (fetch first)
error: failed to push some refs to 'github.com:Mixolyde/blaseballstatus.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

I tried to do a git pull and a git fetch --all before and after running peanut, but that didn't help.

**Comment by kevmoo** [comment]
See https://github.com/kevmoo/peanut.dart#git-tricks

`git update-ref refs/heads/gh-pages origin/gh-pages`

**Comment by Mixolyde** [comment]
Read that, tried to and got

fatal: origin/gh-pages: not a valid SHA1

**Comment by kevmoo** [comment]
Did you try fetching all first?

**Comment by Mixolyde** [comment]
This is my current deploy script:
```
git config --global user.email "mixolyde@gmail.com"
git config --global user.name "Brian Grey"
git config --global push.default "simple"
git fetch --all
peanut
git checkout HEAD -- pubspec.lock
git checkout gh-pages
git pull
git push --set-upstream origin gh-pages
```

And the git pull currently errors out with:
'fatal: refusing to merge unrelated histories'

So bizarre

**Comment by kevmoo** [comment]
Try this...

```
git config --global user.email "mixolyde@gmail.com"
git config --global user.name "Brian Grey"
git config --global push.default "simple"
git fetch --all
git update-ref refs/heads/gh-pages origin/gh-pages
peanut
git push origin gh-pages:gh-pages
```

**Comment by Mixolyde** [comment]
The line:
`git update-ref refs/heads/gh-pages origin/gh-pages`
Still gives:
`fatal: origin/gh-pages: not a valid SHA1`

**Comment by kevmoo** [comment]
Ha! Uh...

can you run `git branch -a` after fetch all – what does it show?

**Comment by Mixolyde** [comment]
```
* (HEAD detached at 9aa7691)
master
remotes/origin/HEAD -> origin/master
remotes/origin/master

```

Also, thank you so much for your time, I super appreciate it!

**Comment by kevmoo** [comment]
```
git config --global user.email "mixolyde@gmail.com"
git config --global user.name "Brian Grey"
git config --global push.default "simple"
git fetch --all
peanut
git push origin gh-pages:gh-pages
```

What's the output w/ this?

**Comment by Mixolyde** [comment]
back to:
```
To github.com:Mixolyde/blaseballstatus.git
! [rejected]        gh-pages -> gh-pages (fetch first)
error: failed to push some refs to 'github.com:Mixolyde/blaseballstatus.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```
This is the end of the peanut output:
```
Branch "gh-pages" was updated with commit 0a8cf014cd
Built web

Branch: HEAD
Commit: 9aa76919eb52baf4845bc729806b432283bb4df8 (dirty)

package:peanut 3.5.1
```
git restore pubspec.lock
git status
```
HEAD detached at 9aa7691
Untracked files:
(use "git add <file>..." to include in what will be committed)
	dartsdk-linux-x64-release.zip

nothing added to commit but untracked files present (use "git add" to track)
```
git pull
`Already up to date.`
git fetch --all
`Fetching origin`
And the above error after
git push origin gh-pages:gh-pages

**Comment by Mixolyde** [comment]
Which CI do you use for dart/peanut project? Circle? Travis?

**Comment by Mixolyde** [comment]
I got it working, hooray! Tech support for codeship pointed me to this gist: https://gist.github.com/mlocher/6124002a5fb693ca2ea1
Between that, your help, and a lot of git status/branch -a commands, I figured out that for whatever reason `fetch --all` and the other fetch commands weren't actually pulling the remote copy locally. I finally ended up with this:
```
git config --global user.email "mixolyde@gmail.com"
git config --global user.name "Brian Grey"
git config --global push.default "simple"
REMOTE="origin"
BRANCH="gh-pages"
git fetch --unshallow || true
git fetch "${REMOTE}" "+refs/heads/*:refs/remotes/${REMOTE}/*"
git fetch "${REMOTE}" "${BRANCH}:${BRANCH}"
peanut
git push "${REMOTE}" "${BRANCH}"
```
The critical one being the `git fetch "${REMOTE}" "${BRANCH}:${BRANCH}"` which actually forced it to fetch a local copy of gh-pages, which peanut then updated and committed to. I think previously peanut was trying to create a 'new' gh-page branch and then couldn't merge it with the remote gh-pages already on github.

Thank you so much for all of your help. You have now contributed to the cultural event of Blaseball!!! I will definitely add you to my acknowledgements list.

https://mixolyde.github.io/blaseballstatus/
